src/checkers/cwe_476.ml

```diff
@@ -1,47 +1,280 @@
+              State.add state variable (Term.tid block)
+          )
+      else
+        state
+
+(** updates a block analysis.
+    The strict call policy decides the behaviour on call and return instructions:
+    strict: all unchecked values get flagged as cwe-hits
+    non-strict: the state gets cleared, it is assumed that the target of the call/return
+    instruction checks all remaining unchecked values. *)
+let update_block_analysis block register_state ~cwe_hits ~function_names ~program ~strict_call_policy =
+  let elements = Blk.elts block in
+  let register_state = Seq.fold elements ~init:register_state ~f:(fun state element ->
+      match element with
+      | `Def def -> update_state_def def state ~cwe_hits
+      | `Phi phi -> state (* We ignore phi terms for this analysis. *)
+      | `Jmp jmp -> update_state_jmp jmp state ~cwe_hits ~function_names ~program ~block ~strict_call_policy
+    ) in
+  State.remove_virtual_registers register_state (* virtual registers should not be accessed outside of the block where they are defined. *)
+
+let print_hit tid ~sub ~function_names ~tid_map =
+  let block = Option.value_exn (Term.find blk_t sub tid) in
+  let jmps = Term.enum jmp_t block in
+  let _ = Seq.find_exn jmps ~f:(fun jmp ->
+      match Jmp.kind jmp with
+      | Call(call) -> begin
+          match Call.target call with
+          | Direct(call_tid) -> Option.is_some (List.find function_names ~f:(fun name ->
+              if name = (Tid.name call_tid) then begin
+                Log_utils.warn "[%s] {%s} (NULL Pointer Dereference) There is no check if the return value is NULL at %s (%s)."
+                  name
+                  version
+                  (Address_translation.translate_tid_to_assembler_address_string tid tid_map)
+                  name;
+                true
+              end else
+                false
+            ))
+          | _ -> false
+        end
+      | _ -> false
+    ) in ()
+
+let check_cwe prog proj tid_map symbol_pairs =
+  let strict_call_policy = true in
+  let symbols = match symbol_pairs with
+    | hd :: _ -> hd
+    | _ -> failwith "[CWE476] symbol_names not as expected" in
+  let function_names = List.map symbols ~f:(fun symb -> "@" ^ symb)  in
+  let subfunctions = Term.enum sub_t prog in
+  Seq.iter subfunctions ~f:(fun subfn ->
+      let cfg = Sub.to_cfg subfn in
+      let cwe_hits = ref [] in
+      let empty = Map.empty Graphs.Ir.Node.comparator in
+      let init = Graphlib.Std.Solution.create empty [] in
+      let equal = State.equal in
+      let merge = State.union in
+      let f = (fun node state ->
+          let block = Graphs.Ir.Node.label node in
+          update_block_analysis block state ~cwe_hits ~function_names ~program:prog ~strict_call_policy
+        ) in
+      let _ = Graphlib.Std.Graphlib.fixpoint (module Graphs.Ir) cfg ~steps:100 ~rev:false ~init:init ~equal:equal ~merge:merge ~f:f in
```

Comment:
100 is a magic number, should it be configurable? Should we set to some lower threshold?

Reply:
I made it configurable. But it is still a magic number. I suppose that 20 could also be enough, but higher numbers are safer to prevent false negatives.